### PR TITLE
ICU-22889 Implemented a recursion limit in the RBNF parsing code to m…

### PR DESCRIPTION
**Pull request by richgillam**
…atch the one already present in the RBNF formatting code.

##### Checklist

- [x] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22889
- [x] Required: The PR title must be prefixed with a JIRA Issue number. 
- [x] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [x] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [x] Issue accepted (done by Technical Committee after discussion)
- [x] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

File: icu4c/source/test/intltest/itrbnf.cpp [hunk 1]
```diff
@@ -2605,6 +2606,36 @@ IntlTestRBNF::TestNumberingSystem() {
...
+        RuleBasedNumberFormat rbnf(badRules[i], parseErr, err);
+        
+        if (U_SUCCESS(err)) {
+            UnicodeString result;
+            rbnf.format(5, result);
+            // we don't actually care about the result; we just want to make sure the function returns
+            
+            Formattable pResult;
+            rbnf.parse("foo", pResult, err);
+            // we don't actually care about the result; we just want to make sure the function returns
```
Comment (grhoten) [inline icu4c/source/test/intltest/itrbnf.cpp:2630]: Please make sure that it actually returns an error for both of these. I think the format fails with a recursion limit being hit. It shouldn't succeed.

File: icu4j/main/common_tests/src/test/java/com/ibm/icu/dev/test/format/RbnfTest.java [hunk 2]
```diff
@@ -1908,4 +1908,34 @@ public void TestNumberingSystem() {
...
+            try {
+                rbnf.format(5);
+                // we don't actually care about the result; we just want to make sure the function returns
+            } catch (IllegalStateException e) {
+                // we're supposed to get an IllegalStateException; swallow it and continue
+            }
+
+            try {
+                rbnf.parse("foo");
+                // we don't actually care about the result; we just want to make sure the function returns
```
Comment (grhoten) [inline icu4j/main/common_tests/src/test/java/com/ibm/icu/dev/test/format/RbnfTest.java:1932]: Please make sure that it actually returns an error for both of these. I think the format fails with a recursion limit being hit. It shouldn't succeed.

**Review by grhoten — CHANGES_REQUESTED** [review]

**Comment by markusicu** [comment]
FWIW looks plausible modulo George's comments.

**Comment by richgillam** [comment]
@grhoten Your concerns are reasonable.  Does this version help?

File: icu4c/source/test/intltest/itrbnf.cpp [hunk 3]
```diff
@@ -2605,6 +2606,37 @@ IntlTestRBNF::TestNumberingSystem() {
...
+        
+        if (U_SUCCESS(err)) {
+            UnicodeString result;
+            rbnf.format(5, result);
+            // we don't actually care about the result and the function doesn't return an error code;
+            // we just want to make sure the function returns
+            
+            Formattable pResult;
+            rbnf.parse("foo", pResult, err);
+            assertTrue("rbnf.parse() didn't return U_INVALID_FORMAT_ERROR!", err = U_INVALID_FORMAT_ERROR);
```
Comment (grhoten) [inline icu4c/source/test/intltest/itrbnf.cpp:2631]: I think you meant `err == U_INVALID_FORMAT_ERROR` and not `err = U_INVALID_FORMAT_ERROR`

File: icu4j/main/common_tests/src/test/java/com/ibm/icu/dev/test/format/RbnfTest.java [hunk 4]
```diff
@@ -1908,4 +1908,34 @@ public void TestNumberingSystem() {
...
+            try {
+                rbnf.format(5);
+                // we don't actually care about the result; we just want to make sure the function returns
+            } catch (IllegalStateException e) {
+                // we're supposed to get an IllegalStateException; swallow it and continue
+            }
+
+            try {
+                rbnf.parse("foo");
+                errln("Parse test didn't throw an exception!");
```
Comment (grhoten) [inline icu4j/main/common_tests/src/test/java/com/ibm/icu/dev/test/format/RbnfTest.java:1932]: Looks good

File: icu4c/source/test/intltest/itrbnf.cpp [hunk 5]
```diff
@@ -2605,6 +2606,37 @@ IntlTestRBNF::TestNumberingSystem() {
     }
 }
 
+void
+IntlTestRBNF::TestInfiniteRecursion() {
+    UnicodeString badRules[] = {
+        ">>",
```
Comment (FrankYFTang) [inline icu4c/source/test/intltest/itrbnf.cpp:2612]: could we also add the case of ">>>" and "<<<". thanks

File: icu4j/main/common_tests/src/test/java/com/ibm/icu/dev/test/format/RbnfTest.java [hunk 6]
```diff
@@ -1908,4 +1908,34 @@ public void TestNumberingSystem() {
         rbnf.setDefaultRuleSet("%ethiopic");
         assertEquals("Wrong result with Ethiopic rule set", "፻፳፫", rbnf.format(123));
     }
+
+    @Test
+    public void TestInfiniteRecursion() {
+        final String[] badRules = {
+                ">>",
+                "<<",
+                "%foo: x=%foo=",
```
Comment (FrankYFTang) [inline icu4j/main/common_tests/src/test/java/com/ibm/icu/dev/test/format/RbnfTest.java:1917]: could we also add the case of ">>>" and "<<<". thanks

File: icu4c/source/test/intltest/itrbnf.cpp (hunk 3 above)
Comment (richgillam) [reply icu4c/source/test/intltest/itrbnf.cpp:2631]: (blush) Good catch!  Sorry about that...

**Review by grhoten — DISMISSED** [review]

File: icu4c/source/test/intltest/itrbnf.cpp (hunk 5 above)
Comment (richgillam) [reply icu4c/source/test/intltest/itrbnf.cpp:2612]: Done!

**Review by FrankYFTang — APPROVED** [review]
LGTM

**Review by FrankYFTang — APPROVED** [review]
LGTM

**Review by grhoten — APPROVED** [review]

**Pull request merged**